### ISSUE TITLE
Discussion on Semantics of NodeCollection and Connect()

### ISSUE THREAD

**Issue by Silmathoron** [nest/nest-simulator#1570]
Following a discussion with @heplesser I would like to discuss possibly overlooked issues with current implementation of ``NodeCollection``.

The ordered nature of ``NodeCollection`` has significant impacts in terms of slicing and connections capabilities.
1. **slicing:** a first issue (see also #1563) is that array slicing must only be done in an ordered way
2. **one-to-one connect** a second issue is that, using ordered containers, the ability to use ``"one_to_one"`` conn_spec becomes extremely limited (see also discussion in #1539) and most use cases would actually require users *not to use* ``NodeCollection`` objects but use lists or arrays instead.

@hakonsbm @stinebuu how important would you say that the ordered nature of ``NodeCollection`` is? Would it be possible to relax this requirement? Could negative steps help?

Relaxing the ordering requirement would restore the full capability of ``"one_to_one"`` connect and solve the slicing issue at the same time, but I don't know whether some internal features rely on the ordered nature of ``NodeCollection`` objects.

This could have major consequences on the implementation that should be chosen for #1563 so I suggest blocking the PR until this issue is resolved.

**EDIT:** thinking about it, I realized that, if we drop the ordering feature, we may also be able to drop the uniqueness requirement (unless I missed something)

**Comment by heplesser** [comment]
## About this note

This issue and related issues and PRs (#1563, #1536, #1539) touch on the principles of how network nodes are managed in scripts for the NEST simulator. In this note, I will try to summarize key considerations in the hope to clarify pertinent aspects and to evolve towards a robust solution.

## General considerations

### Nodes and populations

Network nodes in NEST are always part of a *population*, which may be a single-element population. This population concept exists similarly in other simulation tools. Any given node belongs to exactly one population and each node is unique within its population. For management purposes, each node is assigned a unique ID.

### Nodes have no identities

Historically, the focus of modeling with NEST has been on network models composed of neuron populations connected according to high-level connection rules prescribing how to select source and target neurons for each individual connection when building a *projection* from a source population to a target population. PyNN espouses a similar approach. The classical example of such a network is the Brunel model, but many more and more complex models exist. Ideally, new rules for generating projections should be integrated into NEST as new ConnectionBuilder classes to retain a high-level representation of network connectivity.

As long as connectivity is entirely created in form of projections between populations specified by high-level rules, individual nodes in a population have *no identity* in the sense that each node is equal to all others when considered for connections. Thus, nodes in a population are *principally unordered*, and the node IDs assigned to them are ephemeral: a technical detail for the purpose of administration, but of no consequence to the network model as such. Any relabeling scheme could be applied without affecting network behaviour, and one could even relabel each time a new projection is created (specific realisations of spike trains will vary, but that is not the point here). This even holds true for nodes in populations with spatial placement: the spatial information is encoded in population metadata, which imbues each node with spatial properties, but as long as the mapping between node IDs and spatial information is maintained, any relabeling is possible.

### Node identity imbued by externally prescribed connectivity

In numerous situations, the high-level connection rules provided by NEST are insufficient, either because one would like to explore are rule not yet implemented in NEST but easily available applicable with the help of external tools, detailed connectivity obtained from tools too complex for integration into NEST or when reproducing explicitly specified connectivity. In these cases, once NEST has informed the external tool about the IDs of the nodes in source and target populations, the identities of these nodes are fixed: The external tool will map NEST node IDs to its internal node identifiers before returning information about which connections to create. If the external tool provides such connection information in several batches, node mapping between NEST and the external tool needs to remain fixed to ensure consistent connectivity creation. The mapping thus imbues nodes in NEST with identity: Nodes in NEST cannot be relabeled at will.

### Node identity and devices

Stimulation and recording devices have a special role in a network model. Indeed, some simulation tools do not treat such devices as normal network components at all, in contrast to NEST. In many cases, devices can be considered single-element populations in NEST. The Brunel network, e.g., logically consists of one Poisson-train injecting device per neuron, but in practice this is implemented as a single `poisson_generator`. Similarly, one typically creates on `spike_detector` per population and connects all or a subset of neurons in the population to it. As single-element populations, these device nodes have identity in a sense.

This identity becomes even more prominent when devices are configured and connected using external tools, such as the NEST Instrumentation App. In this case, neurons will also acquire identities, since the graphical selection of neurons for recordings implies a NEST-external mapping of node identifiers.


### Working with nodes

When working with, nodes are accessed in the following situations:
1. to query the parameters and state of nodes,
1. to change the parameters or state of nodes,
1. to query existing connections,
1. to connect nodes with each other, specifically
    a. to connect neurons with neurons,
    a. to connect neurons with devices and vice versa.

For cases 1–3, we only need to be able to specify the set (in the mathematical sense) of nodes which we would like to query or change. For query and change operations, some manner of mapping data returned or values to set with specific nodes may be required, but this is at this level an implementation issue. An simple solution is ordering by node IDs.

For case 4, we need to distinguish between models with and without node identity (imbued by use of external connectivity tools). For models without node identity, where connectivity is constructed exclusively by applying high-level connection rules to populations, we again only need to specify the sets of source and target nodes: each node occurs once in such a set and ordering is irrelevant. 

Where connectivity is prescribed by an external tool, the situation is different: The tool will provide a table specifying pairs of source and target neuron IDs, which could also be seen as a multiset of source-target-ID pairs, since only the mapping of source to target IDs matters, not the relative ordering of the connections among each other (multiset because multiple connections between the same source-target pair are possible; the pairs may be extended to tuples including further synapse parameters generated externally).

Connecting to or from devices raises no principle additional issues; frequently, the device will be a single-node population.


## Consequences for implementation

### Representing populations: NodeConnections

In NEST 3, populations as generated by `nest.Create()` are represented by `NodeCollection` objects (NCs; see also #1375 for further proposal concerning population management). For networks without node identity, populations are sets and NCs are thus a representation of a set. Since iteration over, random selection from and check of membership in NCs is a frequent operation in connection construction, NCs need to be implemented so that these operations are fast. Since `Create()` generates a set of nodes with a contiguous sequence of IDs, a range-based implementation has been chosen. 

NCs can also represent subsets of populations by slicing or combinations of (sub-)populations by union, so that query, change or connection operations can be performed on subpopulations or multiple populations at once. This does not violate the principle that each node belongs to exactly one population: Populations are defined as the result of an individual `Create()` call, all other NCs are representations. The important point is that also subsets and unions of NCs are again, logically speaking, *sets*.

### Indexing of NodeCollections

NEST 3 supports indexing of NCs, e.g.
```python
p = nest.Create('iaf_psc_alpha', n=10)
n5 = p[5]
n137 = p[[1, 3, 7]]
```
This raises one principle and two practical questions: 
1. How can we at all define indexing on a set?
1. If we support indexing, should we allow unordered sequences of indices?
1. If we allow unordered sequences of indices, how do we interpret them?

#### Indexing on a set 
The first questions can be addressed by considering an NC as a [well-ordered set](https://en.wikipedia.org/wiki/Well-order), where the *order is defined by the node IDs* of the elements of the set. By tying the ordering to the node IDs, ordering is well-defined also for unions of NCs.

#### Allowing unordered index sequences
The second question concerns user expectations. If a user perceives an NC essentially as a list, he will expect a different result of the following two lines
```python
p[[1, 3, 7]]
p[[7, 3, 1]]
```
but if she perceives an NC as a set, she will expect the same result from both lines: a set ordered by node ID.  If we are afraid of users falling into the "list trap", we may want to prohibit unordered index lists, but if we properly manage user expectations through documentation, examples (and maybe a name change to `NodeSet`?), unordered indices would provide more flexibility.

#### Interpretation of unordered index sequences
If we define NCs to be well-ordered sets, then this ordering defines the result of indexing with unordered sequences: the result will be a well-ordered set. The only remaining question is whether to permit duplicates in the index sequence. Here, it may be advisable to not permit duplicates in order to robustly meet the user expectation that a sequence of n indicies will return an NC with n elements.

### One-to-one connections

As pointed out by @Silmathoron, the scope of connections that can be created by connecting a source and target nodes specified by NCs using the one-to-one rule is extremely limited is the NCs are well-ordered sets. But in view of the general considerations above, any set of connections created by explicitly specifying pairs of source and target nodes, is a case of externally prescribed connectivity, even if the "external tool" here should be direct user input or a calculation in the Python script controlling the simulation. Thus, one-to-one connections should be considered externally prescribed connections in all cases.

### Externally prescribed connections

Externally prescribed connections are specified by a collection of source-target pairs (or source-target-weight-delay-... tuples). Technically, such a collection could be input in a wide range of ways:
- as a NumPy array with named columns or a Pandas dataframe
- as an adaptor to a file
- as a database accessor
- as an iterator interfacing with an external connectivity generating tool.
Over time, we should attempt to support such breadth, e.g., through the ConnectionGenerator interface. 

At present, the focus should be on connections specified through lists or arrays. Such connections can be specified with a `Connect()` call accepting one source and one target array (or list) of NodeIDs of equal size and a connection rule making explicit that one in this case specifies connections explicitly, not rule-based. Specifying the sources and targets separately keeps notation closer to general `Connect()` syntax. The following names could be choices for the connection rule:
- `explicit`
- `one_to_one`
- `data`
- `prescribed`
To make clear the change from NEST 2, `one_to_one` might not be a good choice, and we may want to consider to retire that rule entirely in NEST 3.

### One-to-many and many-to-one connections

One special situation, especially when connecting to/from devices, is connecting a group of nodes to/from a single node. While in many cases the group will be given by an NC and an all-to-all or fixed-in/outdegree rule used, in some cases one may want to specify the nodes in the group explicitly, e.g., when they are chosen with a graphical tools such as the NEST Instrumentation App. In these cases, no duplicates would be expected in the group.

For this case, we could add the following rule that if `Connect()` is called with either source or target list or array (without duplicates) and the other argument is a single-element population, then the all-to-all rule is applied (as all-to-one or one-to-all).

## Summary / Proposal

1. NodeCollections are well-ordered sets and treated as such.
1. Unordered index sequences are allowed as long as they do not contain duplicates; the result is a well-ordered Node Collection.
1. We abolish the `one_to_one` rule for connections between NodeCollections.
1. A new `explicit` or `data` rule implements explicitly specified connections with input provided by arrays or lists.
1. In combination with a single-element NodeCollection, an array or list input leads to one-to-all or all-to-one connection construction.

**Comment by Silmathoron** [comment]
To complement the proposal of @heplesser on ``NodeCollection``, here are my views on this matter.

## Proposal 1: ``NodeCollection`` as container
If it is technically feasible, we could relax both ordering and uniqueness conditions for ``NodeCollection``. In that case, except for the compact nature of NC, there would be no conceptual difference between this object and an array.
All slicing would now be possible and this would restore the normal use of ``"one_to_one"`` connection as in NEST 2.
Because there are no differences anymore, all ``Connect`` calls would accept any container type.

## Proposal 2: ordered and unique ``NodeCollection``
If we decide that the ordered nature of ``NodeCollection`` is useful/necessary, then I propose the following:
* we remove ``"one_to_one"`` rule in favor of ``"data"`` or ``"explicit"``
* we prohibit array slicing of NC with unordered elements to avoid potential errors that would be extremely difficult to track
* since all remaining ``Connect`` rules from NEST 2 ("all_to_all", "fixed_*degree/number", "*_bernouilli") do not rely on the order of the nodes within the container, we allow the use of any container as an input to ``Connect`` as long as it contains only unique elements
* for the "data" ``Connect``, all input containers are allowed as long as they have the same length.

**Comment by diesmann** [comment]
I am wondering whether with a too clean mathematical definition we 
are not das Kind mit dem Bade ausschuetten.

In Connect(A,B,alg)

we have full flexibility if A and B are just arrays of neuron ids
where any ordering or duplicates are allowed and potentially influcence 
the structure of the resulting network.

This interpretation has served us well for a long time.

The disadvantage of this is that we often have extremely boring 
arrays of the sort 1,2,3,...n. Although there is little information
in this array, a naive representation would occupy n memory elements.

A way out of this is to allow in the place of native arrays also 
generators. These are objects which are representing sequences of 
ids in a compact way and deliver the next id upon the next call.
The concept is known from SLI and Python.
The sequences can be from a staring integer to an ending one, in
steps of 1 but also with larger steps, and in particular also 
negative steps. In Theory even arbitrary integer functions are possible.

I always viewed NodeCollections as implementations of the concept
described in the last paragraph.

If an id structure does not fit into the available compressed 
representation, one can always fall back to a plain array.

**Comment by heplesser** [comment]
The [NEST Open Developer VC today](https://github.com/nest/nest-simulator/wiki/2020-05-11-Open-NEST-Developer-Video-Conference) supported [@Silmathoron's Proposal 2](https://github.com/nest/nest-simulator/issues/1570#issuecomment-623921758). The proposal is now as follows:

1. `NodeCollection`s are well-ordered sets
    1. Each NodeID occurs at most once in an NC
    1. Iteration over NCs is in order of increasing node ID
    1. NCs can only be sliced with positive step
    1. NCs can only be indexed with ordered sequences of indices

1. `Connect()` accepts definite-length iterables compatible with connection rules
    1. The following specifications relate to the `pre` and `post` arguments of `Connect()`
    1. The arguments can be any iterable that knows its length, i.e., for which `len()` returns a finite, positive value.
    1. If the connection rule is `'explicit'`, `pre` and `post` must have the same size and will be connected in a one-to-one manner.
    1. Otherwise, if the argument is not an NC, it must not contain duplicate node IDs. It is left to the implementation if uniqueness is checked before the first connection is created.
    1. All values given as node IDs in an iterable must be valid node IDs. It is left to the implementation if validity is checked before the first connection is created.

### Examples
#### Slicing of NodeCollection
```python
nodes = nest.Create(`iaf_psc_exp`, n=10)
nodes[[1, 5, 8]]

NodeCollection(metadata=None,
               model=iaf_psc_exp, size=1, first=2;
               model=iaf_psc_exp, size=1, first=6;
               model=iaf_psc_exp, size=1, first=9)
```
**Note:** Indices `[1, 5, 8]` are zero-based, while node IDs are one-based.

```python
nodes[[1, 2, 2]]      # ERROR: duplicate index

nodes[[2, 1, 3]]     # ERROR: indices not in order
```

#### Connect
```python
nodes = nest.Create(`iaf_psc_exp`, n=10)

# default: all-to-all connection
nest.Connect(nodes, nodes)   

# one argument iterable, one node collection 
nest.Connect([1, 7, 4], nodes, conn_spec={'rule': 'pairwise_bernoulli', 'p': 0.4})

# ERROR: iterable contains non-unique node ID
nest.Connect([1, 4, 4], nodes, conn_spec={'rule': 'pairwise_bernoulli', 'p': 0.4})

# ERROR: iterable contains non-existing node ID
nest.Connect([1, 4, 44], nodes, conn_spec={'rule': 'pairwise_bernoulli', 'p': 0.4})

# explicit rule: one-to-one connection 
nest.Connect(nodes, nodes, conn_spec='explicit')

# explicit rule with lists: one-to-one connection, repetitions possible
nest.Connect([1, 5, 3, 7, 3], [1, 2, 1, 2, 1], conn_spec='explicit')

# explicit rule with NodeConnection and list: one-to-one connection, repetitions possible in list
nest.Connect(nodes[:5], [1, 2, 1, 2, 1], conn_spec='explicit')

# ERROR: pre/post with repetitions or unordered and other rule than `explicit`
nest.Connect([1, 2, 3, 3], [4, 3, 2, 1], conn_spec={'rule': 'fixed_indegree', 'indegree': 3})

**Comment by Silmathoron** [comment]
I'm fine with the last proposal.
I think the only thing that needs a formal decision is the name of the new rule? Is "explicit" explicit enough? Is "data" better? Should we look for yet another rule? Should we reuse "one_to_one" there?

Once that is settled, how shall we proceed for the implementation?
I think this is compatible with #1539 so one question is as follow:
* do we first make a "good enough" implementation with the current conversions to NC and numpy arrays, then make an issue to improve the "non-standard cases" (minimizing the copies, etc)? (I think this would be #1539 more or less as it is now)
* do we first think about how to implement this as efficiently as possible (i.e. seeing whether we can move arrays directly down to the C++ level without copying them and how we handle the four different cases there: NC/NC, arr/arr, arr/NC, NC/arr)?

@hakonsbm any opinion on that?

Second, in what order do we do things?
I personally think the changes are broadly independent so we should be able to do in (almost) any order:
- [ ] forbid unordered slices of NC
- [ ] remove "one_to_one" in favor of the new rule
- [ ] make ``Connect`` accept the different inputs

The only thing is that it requires fewer changes to change ``Connect`` after we have decided on a name for the new rule.

If we go for the "good enough" implementation first, I would suggest to chose the new rule name and change ``Connect`` in a single PR, as it is currently done in #1539.

**Comment by BSVogler** [comment]
I welcome this design.

Is _explicit_ the same as _one_to_one_? I think _one_to_one_ is more self-explaining than explicit.

**Comment by Silmathoron** [comment]
@BSVogler thank you for your feedback!
Yes, list have to be used if the user wants duplicates, but this would only occur with the explicit/one-to-one rule.
As to which name is better, I must say that I agree with you that ``"one_to_one"`` might be more self-explanatory so keeping it might be a good idea...
@heplesser @diesmann should we make a poll to see which name is most intuitive for the NEST community?
If so could we list some pros and cons first?

I must say I don't really see any cons to keeping ``"one_to_one"``...
This would limit the number of lines we would break in user scripts and it would probably be the most intuitive rule name since users already know it.

**Comment by hakonsbm** [comment]
> * do we first make a "good enough" implementation with the current conversions to NC and numpy arrays, then make an issue to improve the "non-standard cases" (minimizing the copies, etc)? (I think this would be #1539 more or less as it is now)
> 
> * do we first think about how to implement this as efficiently as possible (i.e. seeing whether we can move arrays directly down to the C++ level without copying them and how we handle the four different cases there: NC/NC, arr/arr, arr/NC, NC/arr)?
> 
> 
> @hakonsbm any opinion on that?

@Silmathoron To be able to use a NumPy array in C++ for connecting without copying it, it would have to be wrapped in a NodeCollection. That means that in addition to the existing primitive and composite NodeCollections, we would need a NumPy array version as well.

I therefore suggest that we start by changing the `Connect()` interface with a "good enough" implementation only, where we always do the conversion to NodeCollections in Python if the `'explicit'`/`'one_to_one'` rule is not used.

**Comment by Silmathoron** [comment]
@hakonsbm sounds good to me!

@heplesser @diesmann how should we proceed for the choice of the connectivity rule?

**Comment by heplesser** [comment]
@Silmathoron @BSVogler `one_to_one` is indeed more descriptive than `explicit` or `data`. My main reason for suggesting a change was that this would be the one rule that has different semantics (allows unordered and repetitions). In NEST2, all connection rules admit all arguments (just lists), now we get one that is special. Therefore, I thought that a name change compared to NEST2 would make sense.

On the other hand, any NEST2 script that uses `one_to_one` will work in exactly the same way in NEST3, so there is no need to change names to break existing scripts and protect users from a situation where the same code suddenly does different things. And for all new users of NEST (which in the long run will be the majority), naming history is irrelevant anyways.

So I think keeping `one_to_one` is fine. We could have one last confirmation on this in the Developer VC on 25 May. Until then, I think we should proceed with #1539 as you [suggested above](https://github.com/nest/nest-simulator/issues/1570#issuecomment-627377510).

In my [proposal post](https://github.com/nest/nest-simulator/issues/1570#issuecomment-626916028) in items 2.iv-v, I stated that it should be up to the implementation to check before creating the first connection if all IDs in the pre and post arguments exist and are unique when `Connect()` is called with non-NC arguments and a different connection rule than `one_to_one`.  In our initial implementation we should include these checks upfront to help users sort out logic errors in scripts.

**Comment by Silmathoron** [comment]
@heplesser OK, then I update #1539 some time this week and ping you

**Comment by Silmathoron** [comment]
Hi @heplesser, I lost track of where we were in implementing this... could we summarize where we are with respect to the [todo list](https://github.com/nest/nest-simulator/issues/1570#issuecomment-627377510)?

I think #1539 is ready (I answered @hakonsbm comments)

We blocked issue #1563 so everything would be discussed here. We decided we would raise an error if unordered ids are used to slice a ``NodeCollection`` but I don't know whether this has been implemented yet. Could you post the result of the discussion there and close (or not) depending on the situation?

**Comment by terhorstd** [comment]
@Silmathoron, if your PR is already far enough and no more discussion is required here, I'd go ahead and remove the discussion label. Is there something else to discuss here?

**Comment by Silmathoron** [comment]
@terhorstd yes, the PR is theoretically ready to be merged as soon as it gets a 2nd review and I don't think we have anything else to discuss, so the discussion label can go.
I'm still waiting for somebody to tell me whether we actually took care or the ordering problem, though, so maybe unblock #1563 and write there that an error should be raised?

**Comment by stinebuu** [comment]
As #1536 and #1539 is merged and #1563 is closed, it seems like all related PRs and issues are either merged or closed. I am therefore closing this issue. Thanks to all who contributed!
